refactor(core): initial implementation of `{#defer}` block runtime

AndrewKushnir opened the pull request:
This commit adds an initial implementation of the `{#defer}` block runtime, which supports the `when` conditions. More conditions and basic prefetching support will be added in followup PRs.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
refactor(core): initial implementation of `{#defer}` block runtime

This commit adds an initial implementation of the `{#defer}` block runtime, which supports the `when` conditions. More conditions and basic prefetching support will be added in followup PRs.

JoostK commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -6,48 +6,109 @@
+const _requestIdleCallback =
+    typeof requestIdleCallback !== 'undefined' ? requestIdleCallback : setTimeout;
+const _cancelIdleCallback =
+    typeof cancelIdleCallback !== 'undefined' ? cancelIdleCallback : clearTimeout;
```
This should arguably use the same condition as `requestIdleCallback`, as a `requestIdleCallback`/`setTimeout` must be cancelled using `cancelIdleCallback`/`clearTimeout` respectively.

AndrewKushnir replied:
Thanks for the comment, the code will be updated asap.

AndrewKushnir commented:
[Exploratory presubmit](https://test.corp.google.com/OCL:557707200:BASE:557707226:1692250364394:207e2976).

crisbeto commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -6,48 +6,108 @@
+    loadingPromise: null,
+  });
+
+  templateInternal(index, null, 0, 0, deferBlockConfig);
```
Isn't this called already in the creation block when we define the primary block template?

AndrewKushnir replied:
This is a separate call with the main goal to create an LContainer, where we'll be inserting views based on defer block conditions. We are not reusing other LContainers created for inner blocks, this would minimize the amount of refactoring later when we split `template` instruction into a `block` and a `container` ones.

crisbeto commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -6,48 +6,108 @@
+
+  // Defer block details are needed only once during the first creation pass,
+  // so we wrap an object with defer block details into a function that is only
+  // invoked once to avoid re-constructing the same object for each subsequent
```
I have a feeling that constructing the function would be more expensive than reconstructing the object literal.

alxhub replied:
Given that everything in the `deferBlockConfig` is either static, from constant arguments, or already in the `consts` array, could we put the whole config in the `consts` array as a single object instead?

AndrewKushnir replied:
That's an interesting idea, we can explore this as an optimization in a followup PR.

crisbeto commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -56,10 +116,21 @@ export function ɵɵdeferWhen(value: unknown) {}  // TODO: implement runtime log
+               triggerDeferBlock(lView, tNode);
+               cancelIdleCallback(id);
+             }) as number;
+  storeLViewOnDestroy(lView, () => _cancelIdleCallback(id));
```
Do we also want to clear this callback from memory once the block is done loading? At that point we won't have to keep it in memory anymore.

crisbeto commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -133,3 +204,256 @@ export function ɵɵdeferOnViewport(target?: unknown) {}  // TODO: implement run
+    // represents a `{#defer}` block, so always refer to the first one.
+    const viewIndex = 0;
+    removeLViewFromLContainer(lContainer, viewIndex);
+    const embeddedLView = createAndRenderEmbeddedLView(hostLView, tNode, {});
```
Could we pass `null` as the object here instead of the empty object?

crisbeto commented on packages/core/src/render3/interfaces/container.ts:
```diff
@@ -55,7 +56,38 @@ export const DEHYDRATED_VIEWS = 10;
+ * would be added later to keep track of `after` and `maximum` features
+ * (which would require per-instance state).
+ */
+export interface LDeferBlockDetails {
```
Do we want to follow the pattern of using an array literal for storing information in the L structures?

crisbeto commented on packages/core/test/acceptance/defer_spec.ts:
```diff
@@ -0,0 +1,486 @@
+import {ɵsetEnabledBlockTypes as setEnabledBlockTypes} from '@angular/compiler/src/jit_compiler_facade';
+import {Component, Input, QueryList, ViewChildren, ɵDEFER_BLOCK_DEPENDENCY_INTERCEPTOR} from '@angular/core';
+import {TestBed} from '@angular/core/testing';
+
```
It looks like all the tests are using standalone components. Should we have some with an `NgModule` to ensure that everything works correctly there?

AndrewKushnir replied:
Thanks for the comment. We'd need to review the DX around NgModule-declared components to see which patterns we can or can not support (in which case we'd need to show a clear error message).

crisbeto commented on packages/core/test/acceptance/defer_spec.ts:
```diff
@@ -0,0 +1,486 @@
+import {ɵsetEnabledBlockTypes as setEnabledBlockTypes} from '@angular/compiler/src/jit_compiler_facade';
+import {Component, Input, QueryList, ViewChildren, ɵDEFER_BLOCK_DEPENDENCY_INTERCEPTOR} from '@angular/core';
+import {TestBed} from '@angular/core/testing';
+
```
Should we have a test with nested `defer` blocks?

AndrewKushnir replied:
There are tests for nested defer blocks in the spec file (last `describe` block). Please let me know if you have any specific tests in mind, I can add more.

pkozlowski-opensource commented on packages/core/src/render3/interfaces/container.ts:
```diff
@@ -144,6 +176,12 @@ export interface LContainer extends Array<any> {
+   * If this LContainer represents an instance of a `{#defer}` block -
+   * this field contains instance-specific information about the block.
+   */
+  [DEFER_BLOCK_DETAILS]: LDeferBlockDetails|null;
```
Do we need to store this on the LContainer level? I'm worried that every single container will "pay the price" here. How about allocating another slot in `data`?

pkozlowski-opensource commented on packages/core/src/render3/instructions/template.ts:
```diff
@@ -75,14 +75,26 @@ export function ɵɵtemplate(
+      index, templateFn, decls, vars, tagName, attrsIndex, localRefsIndex, localRefExtractor);
+}
+
+export function templateInternal(
```
I would probably do this refactoring only when we do the block / container split.

pkozlowski-opensource commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -858,6 +859,88 @@ export interface TProjectionNode extends TNode {
   value: null;
 }
 
+/**
```
nit: do we want to add those interfaces in the existing node.ts file or maybe create a dedicated file for deffered blocks.

AndrewKushnir committed:
fixup! refactor(core): initial implementation of `{#defer}` block runtime

AndrewKushnir commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -6,48 +6,108 @@
+    index: number, primaryTmplIndex: number, dependencyResolverFn?: DependencyResolverFn|null,
+    loadingTmplIndex?: number|null, placeholderTmplIndex?: number|null,
+    errorTmplIndex?: number|null, loadingConfigIndex?: number|null,
+    placeholderConfigIndex?: number|null) {
```
This is similar to how other instructions handle arguments. We can do a followup cleanup across all instructions later once there is a clear pattern on how we want to type them.

AndrewKushnir commented:
@pkozlowski-opensource @crisbeto @alxhub thanks for reviewing this PR! 👍  I've addressed the feedback, mostly by changing the code, but in some cases we can make further improvements as a followup.

There is a relatively big change that was added and which I wanted to highlight: after a discussion with @pkozlowski-opensource offline, I've updated the code to store `defer` block info (static and instance-specific) in `LView` and `TView.data` slightly differently: for each `defer` block we now allocate an extra slot in the `data` part of an `LView` (and `TView.data`) right after the `defer` instruction slot itself. So if we have the following generated code:

```
defer(N, ...);
```
the layout would be:
```
LView[HEADER_OFFSET + N] = LContainer; // container which we use to insert views
LView[HEADER_OFFSET + N + 1] = [DeferBlockInstanceState]; // instance-specific state of a block

TView.data[HEADER_OFFSET + N] = TNode for an LContainer; // *no* extra info about corresponding `defer` block
TView.data[HEADER_OFFSET + N + 1] = { ... }; // static info of this `defer` block
```

Later on, once we have a separate `block` and `container` instructions, the `LView[HEADER_OFFSET + N]` would host the `container`-related info (basically the same info) and this extra `N+1` slot would still contain defer-specific info, so this layout makes it more future-proof. Also, this layout avoids an extra slot in **every** `LContainer`, which was previously dedicated to an instance-specific state of a block.

Please take a look when you get a chance :) 

Thank you.

pkozlowski-opensource commented on packages/core/src/render3/interfaces/type_checks.ts:
```diff
@@ -56,3 +56,7 @@ export function isProjectionTNode(tNode: TNode): boolean {
 }
+
+export function isDestroyed(lView: LView): boolean {
+  return (lView[FLAGS] & LViewFlags.Destroyed) === LViewFlags.Destroyed;
```
Great move 👍 

There are other places in the code doing `lView[FLAGS] & LViewFlags.Destroyed` - we should refactor those to new utility functions

JeanMeche commented on packages/core/src/core_render3_private_export.ts:
```diff
@@ -68,6 +68,8 @@ export {
   setClassMetadataAsync as ɵsetClassMetadataAsync,
   setLocaleId as ɵsetLocaleId,
   store as ɵstore,
+  ɵDeferBlockDependencyInterceptor,
```
This export seems to be unused. Do you expect it to be used in the near future ?

pkozlowski-opensource commented on packages/core/test/acceptance/defer_spec.ts:
```diff
@@ -0,0 +1,486 @@
+  beforeEach(() => setEnabledBlockTypes(['defer']));
+  afterEach(() => setEnabledBlockTypes([]));
+
+  it('should work in simple cases', async () => {
```
nit: how about making this test name in-line with what it does, ex.: "shows defer block after showing placeholder, loading and error states" - or something else. For me it is not clear what "should work" and "simple cases" mean...

AndrewKushnir replied:
Thanks Pawel, I've updated the description.

pkozlowski-opensource commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -6,48 +6,109 @@
+  const oldValue = lView[bindingIndex];
+  // If an old value was `true` - don't enter the path that triggers
+  // defer loading.
+  if (oldValue !== true && bindingUpdated(lView, bindingIndex, value)) {
```
Do we need the intermediate `oldValue` variable here? Wouldn't it be enough to convert incoming `value` to `Boolean`? 

`if (bindingUpdated(lView, bindingIndex, Boolean(value))) {`

AndrewKushnir replied:
We want to guard against changing the binding value from `true` to something else. The `when` value can only go to `true` once, which triggers loading. After that, a defer block does not react to value changes.

pkozlowski-opensource replied:
Couldn't this be handled with sth like `loaded || Boolean(value)`?

pkozlowski-opensource replied:
Other thought: it seems like what we are dealing with here is a state machine and the move between states happen based on the previous states _and_ the incoming value. Maybe it would be possible to store the previous state in the bindings, and use calculated "new state" as the new incoming value?

I'm not fully sure if it will simplify the code, but looking for a way of make it easier to parse here.

AndrewKushnir replied:
Thanks for the feedback. I've changed the logic to rely on an existing block state (vs relying on previous value in a slot).

jessicajaniuk reviewed (approved):
This looks great to me. I think we'll probably want to add more tests for nested defer blocks to ensure the different types of nestings with conditions behave as we expect them to. We can add those as more of the triggers get implemented.

pkozlowski-opensource reviewed (approved):
LGTM!

JoostK commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -6,48 +6,114 @@
+  const bindingIndex = nextBindingIndex();
+  const newValue = !!value;  // handle truthy or falsy values
+
+  if (bindingUpdated(lView, bindingIndex, value)) {
```
Should this be `newValue` instead of `value`? If not, I propose moving `newValue` to be declared within the `if` block.

pkozlowski-opensource replied:
Oh, I think it is a good catch, actually. Would it mean a missing test?
Plus I've got _slight_ preference for using `Boolean(value)` for boolean conversion as it makes the intention clearer, IMO (despite adding more bytes)

AndrewKushnir replied:
Thanks for the feedback, I've updated the code to move the condition inside the `if` block and used `Boolean` instead of `!!`.

AndrewKushnir commented:
[Presubmit](http://test/OCL:557707200:BASE:560721591:1693237424132:e53a810c).

jessicajaniuk commented:
This PR was merged into the repository by commit c4deaac5b0816ebee0cb6f92ba02116bec2ca623.
